examples/scripts/TreeCompareAuto.C

```diff
@@ -0,0 +1,344 @@
+/********************************************************************************
+ *    Copyright (C) 2023 GSI Helmholtzzentrum fuer Schwerionenforschung GmbH    *
+ *                                                                              *
+ *              This software is distributed under the terms of the             *
+ *              GNU Lesser General Public Licence (LGPL) version 3,             *
+ *                  copied verbatim in the file "LICENSE"                       *
+ ********************************************************************************/
+
+// The macro compares all the leaves of two TTree objects.
+// The first and strongest comparison is that the entries in the two leaves
+// to compare  are idential.
+// To check this one loops over the entries in the trees, calculate the
+// difference between the two values and fills the result in a histogram.
+// If all entries are absolutely identical the result in the histogram is a
+// delta function at 0.
+// If the first check for identical entries fails it is checked if there are
+// only some changes in the order of the entries. In this case the two
+// produced histgrams (one for each tree) are identical.
+// The last check which again is only done if both previous test fails uses a
+// Kolmogorov test to check if the produced hostograms are comparable on a
+// statistical base.
+// If at least for one leaf all theree comparisons fail the complete test
+// fails.
+
+#include <RtypesCore.h>
+#include <TCanvas.h>
+#include <TCollection.h>   // for TRangeDynCast
+#include <TFile.h>
+#include <TH1.h>
+#include <TKey.h>
+#include <TLeaf.h>
+#include <TMath.h>
+#include <TString.h>
+#include <TTree.h>
+#include <algorithm> // for std::any_of
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <string_view>
+#include <utility>
+#include <vector>
+
+std::vector<std::string> GetLeafNames(TTree&);
+
+bool operator==(TH1 const&, TH1 const&);
+
+int TreeCompareAuto(TString fileName1 = "", TString fileName2 = "")
+{
+    if (fileName1.IsNull() || fileName2.IsNull()) {
+        cout << "Filenames are not defined" << endl;
+        exit(42);
+    }
+
+    // Get the output tree from the original file
+    std::unique_ptr<TFile> file1{TFile::Open(fileName1, "READ")};
+    if (!file1) {
+        std::cout << "Could not open file " << fileName1 << std::endl;
+        return 42;
+    }
+
+    // Find a tree in the file
+    TList* keylist = file1->GetListOfKeys();
+    TKey* key;
+    TIter keyIterator(keylist);
+    TString treeName;
+    int numTreeInFile{0};
+    while ((key = static_cast<TKey*>(keyIterator()))) {
+        if (key->ReadObj()->InheritsFrom("TTree")) {
+           treeName     = key->GetName();
+           std::cout << "Found TTree with name " << treeName << std::endl;
+           numTreeInFile++;
+        }
```

Comment:
`TKey::ReadObj` returns an owning pointer (see #1461).

Maybe move this in front of the loop?
```cpp
    std::unique_ptr<TTree> tree1;
```

And then:

```suggestion
        tree1.reset(key->ReadObject<TTree>());
        if (tree1) {
           treeName     = key->GetName();
           std::cout << "Found TTree with name " << treeName << std::endl;
           numTreeInFile++;
        }
```